Add capability to use RAP/HRRR MASSDEN as smoke IC/LBCs

jordanschnell opened the pull request:
## DESCRIPTION OF CHANGES: 
Previously, massden/smoke was not part of the variable list to ingest into IC/LBCs from RAP/HRRR models. This PR adds the massden/smoke to the mapping table and adds the necessary functionality to read in and convert to the correct units. An additional allocated array is created when using RAP/HRRR for ICs/LBCs during the smoke read to be able to convert units. Other models (e.g., GFS) will not attempt to read in smoke.

## TESTS CONDUCTED: 

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera, Hercules and WCOSS2). Done using d126cb6.
- [x] Compile branch on Hera using GNU. Done using d126cb6.
- [x] Compile branch in 'Debug' mode on WCOSS2 with no warnings. Done using d126cb6.
- [x] Run unit tests locally on any Tier 1 machine. Done on Hera using d126cb6 (and Gnu). All tests passed.
- [x] Run `chgres_cube` consistency tests locally on all Tier 1 machines. Done using d126cb6. All tests - and the new test - passed.

Additional tests performed.

- Changes based on develop (72fb94fc6bb8c917a457ff8105c1552eb8c02f7a)
- Compiled on CENTOS node on Jet
- Created IC/BCs from RAP for RRFS 05/14/2023 00Z cycle using RAP cycle 05/13/2023 21Z.
- @MichaelLueken - Successfully tested the branch within the SRW App. See https://github.com/ufs-community/UFS_UTILS/pull/923#issuecomment-2394517060
- Doxygen was built with no warnings on Hera using d126cb6.

## DEPENDENCIES:
none

## DOCUMENTATION:
Update readthedocs: https://schnel.readthedocs.io/en/rap_icbc/ufs_utils.html#variable-mapping-varmap-table

## ISSUE: 
Fixes #851.

## CONTRIBUTORS: 
Johana Romero-Alvarez (CIRES/NOAA GSL)


![Screenshot 2024-03-1
<img width="725" alt="Screenshot 2024-03-19 at 6 37 28 PM" src="https://github.com/ufs-community/UFS_UTILS/assets/46031459/2fd276fa-4e8a-481f-9ed2-34c7a35eda50">
9 at 6 35 55 PM](https://github.com/ufs-community/UFS_UTILS/assets/46031459/476431bc-3b61-4a7a-8dd0-d30774592a9a)
![Screenshot 2024-03-19 at 6 39 17 PM](https://github.com/ufs-community/UFS_UTILS/assets/46031459/de46975f-4b8b-4803-8bb3-1bf757cf4606)



<img width="725" alt="Screenshot 2024-03-19 at 6 37 28 PM" src="https://github.com/ufs-community/UFS_UTILS/assets/46031459/8469b4cf-3952-4e16-977a-9acfcb743b36">

jordanschnell committed:
Add capability to ingest RAP/HRRR smoke for IC/BC

jordanschnell committed:
deallocate dummy3d_pres if allocated

jordanschnell committed:
Fix typos

jordanschnell committed:
Explicit read of pressure during smoke portion of loop

jordanschnell committed:
cleanup

jordanschnell committed:
Fix character length in massden

jordanschnell commented:
> I am curious what project or implementation this PR supports.

I'm not sure exactly which details you need, but this is largely research oriented e.g., For computational efficiency, for retrospective experiments prior to RRFS-A SD implementation, we often only want to run the CONUS domain but would still like e.g., Canadian originated smoke to enter the domain (e.g., DRSA FIRE3). The same could be said for using HRRR IC/BCs for the fire nest. This wIll also support NOAA ERB funded work.

GeorgeGayno-NOAA commented:
We ask that developers run the consistency tests located in `./reg_tests/chgres_cube`. I ran them on Hera (CentOS nodes). Two of the tests failed. (The number of tracers in the gfs_ctrl.nc file differ from the baseline). Can you run these tests on Jet? Just invoke `./driver.jet.sh`.

jordanschnell commented:
> We ask that developers run the consistency tests located in `./reg_tests/chgres_cube`. I ran them on Hera (CentOS nodes). Two of the tests failed. (The number of tracers in the gfs_ctrl.nc file differ from the baseline). Can you run these tests on Jet? Just invoke `./driver.jet.sh`.

I ran the tests on Jet and also received the two failures (13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 and 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2). This makes sense though, sense smoke is being added to the files and so e.g., there is an additional tracer in the out files. How would you like this to be addressed?

GeorgeGayno-NOAA commented:
> > We ask that developers run the consistency tests located in `./reg_tests/chgres_cube`. I ran them on Hera (CentOS nodes). Two of the tests failed. (The number of tracers in the gfs_ctrl.nc file differ from the baseline). Can you run these tests on Jet? Just invoke `./driver.jet.sh`.
> 
> I ran the tests on Jet and also received the two failures (13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 and 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2). This makes sense though, since smoke is being added to the files and so e.g., there is an additional tracer in the out files. How would you like this to be addressed?

The number of tracers in `gfs_ctrl.nc` should match those the in the atmospheric 'out' file.

GeorgeGayno-NOAA commented:
As a test, I ran the `13km.conus.rap.grib2` regression test with all the smoke records removed from the input GRIB2 file.

The test script was updated as follows:

```
--- a/reg_tests/chgres_cube/13km.conus.rap.grib2.sh
+++ b/reg_tests/chgres_cube/13km.conus.rap.grib2.sh
@@ -16,9 +16,9 @@ export CRES=778
 export KMRES=13km
 export FIXfv3=${HOMEreg}/fix/RRFS_CONUS_${KMRES}

-export COMIN=${HOMEreg}/input_data/rap.grib2
+export COMIN=/work2/noaa/da/ggayno/save/UFS_UTILS.schnell/data

-export GRIB2_FILE_INPUT=1921221000900
+export GRIB2_FILE_INPUT=rap.grb2
 export VCOORD_FILE=${HOMEufs}/fix/am/global_hyblev.l64.txt
 export VARMAP_FILE=${HOMEufs}/parm/varmap_tables/GSDphys_var_map.txt
```

The modified input data and the script used to create it is here: `/work2/noaa/da/ggayno/save/UFS_UTILS.schnell/data`

GeorgeGayno-NOAA commented:
For the first test using the modified input data, I updated the varmap file as follows:

```
--- a/parm/varmap_tables/GSDphys_var_map.txt
+++ b/parm/varmap_tables/GSDphys_var_map.txt
@@ -25,4 +25,4 @@ icetk               icetk     set_to_fill   265.0     S
 weasd               weasd     set_to_fill   0.0       S
 snod                snod      set_to_fill   0.0       S
 tprcp               tprcp     set_to_fill   0.0       S
-massden             smoke     set_to_fill   1E-12     T
+massden             smoke     stop          1E-12     T
```

Since there are no smoke records, processing should stop. And that is what the log file shows:

```
- CALL FieldGather TEMPERATURE.
 - READ PRESSURE FOR SMOKE CONVERSION.
 - FATAL ERROR:
 READING massden at level      1 hybrid       . TO MAKE THIS NON-FATAL, CHANGE S
 TOP TO SKIP FOR THIS VARIABLE IN YOUR VARMAP FILE.
 - IOSTAT IS:            0
Abort(999) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 999) - process 0

```

GeorgeGayno-NOAA commented:
> > For the first test using the modified input data, I updated the varmap file as follows:
> > ```
> > --- a/parm/varmap_tables/GSDphys_var_map.txt
> > +++ b/parm/varmap_tables/GSDphys_var_map.txt
> > @@ -25,4 +25,4 @@ icetk               icetk     set_to_fill   265.0     S
> >  weasd               weasd     set_to_fill   0.0       S
> >  snod                snod      set_to_fill   0.0       S
> >  tprcp               tprcp     set_to_fill   0.0       S
> > -massden             smoke     set_to_fill   1E-12     T
> > +massden             smoke     stop          1E-12     T
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since there are no smoke records, processing should stop. And that is what the log file shows:
> > ```
> > - CALL FieldGather TEMPERATURE.
> >  - READ PRESSURE FOR SMOKE CONVERSION.
> >  - FATAL ERROR:
> >  READING massden at level      1 hybrid       . TO MAKE THIS NON-FATAL, CHANGE S
> >  TOP TO SKIP FOR THIS VARIABLE IN YOUR VARMAP FILE.
> >  - IOSTAT IS:            0
> > Abort(999) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 999) - process 0
> > ```
> 
> Can you explain why you modified this?

The varmap files have several choices with how to process fields: 

> > For the first test using the modified input data, I updated the varmap file as follows:
> > ```
> > --- a/parm/varmap_tables/GSDphys_var_map.txt
> > +++ b/parm/varmap_tables/GSDphys_var_map.txt
> > @@ -25,4 +25,4 @@ icetk               icetk     set_to_fill   265.0     S
> >  weasd               weasd     set_to_fill   0.0       S
> >  snod                snod      set_to_fill   0.0       S
> >  tprcp               tprcp     set_to_fill   0.0       S
> > -massden             smoke     set_to_fill   1E-12     T
> > +massden             smoke     stop          1E-12     T
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since there are no smoke records, processing should stop. And that is what the log file shows:
> > ```
> > - CALL FieldGather TEMPERATURE.
> >  - READ PRESSURE FOR SMOKE CONVERSION.
> >  - FATAL ERROR:
> >  READING massden at level      1 hybrid       . TO MAKE THIS NON-FATAL, CHANGE S
> >  TOP TO SKIP FOR THIS VARIABLE IN YOUR VARMAP FILE.
> >  - IOSTAT IS:            0
> > Abort(999) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 999) - process 0
> > ```
> 
> Can you explain why you modified this?

According to the README file, users can adjust how fields are processed using the varmap file. (See the instructions for column 3). Options include 'stop', 'skip' 'set_to_fill' etc. These different options need to be tested. So, I am temporarily modifying the varmap file.

https://github.com/ufs-community/UFS_UTILS/blob/develop/parm/varmap_tables/README

GeorgeGayno-NOAA commented:
Next, test the 'skip' function. Modify the varmap table as follows:

```
--- a/parm/varmap_tables/GSDphys_var_map.txt
+++ b/parm/varmap_tables/GSDphys_var_map.txt
@@ -25,4 +25,4 @@ icetk               icetk     set_to_fill   265.0     S
 weasd               weasd     set_to_fill   0.0       S
 snod                snod      set_to_fill   0.0       S
 tprcp               tprcp     set_to_fill   0.0       S
-massden             smoke     set_to_fill   1E-12     T
+massden             smoke     skip          1E-12     T
```

The log file shows:

 ```
WARNING: SKIPPING massden IN FILE
 WARNING: SKIPPING massden IN FILE
 WARNING: SKIPPING massden IN FILE
```

And the out.atm.tile7.nc file contains a smoke record. Because the input GRIB2 file does not contain smoke, that was unexpected.  According to the README, when 'skip' is used, there should not be a smoke record.

```
Herc /work2/noaa/stmp/ggayno/reg-tests/chgres-cube.skip/13km_conus_rap_grib2\ $ ncdump -h out.atm.tile7.nc
        float ice_aero(lev, lat, lon) ;
                ice_aero:coordinates = "geolon geolat" ;
        float sgs_tke(lev, lat, lon) ;
                sgs_tke:coordinates = "geolon geolat" ;
        float smoke(lev, lat, lon) ;
                smoke:coordinates = "geolon geolat" ;
        float u_w(lev, lat, lonp) ;
                u_w:coordinates = "geolon_w geolat_w" ;
```

@LarissaReames-NOAA and @JeffBeck-NOAA - you understand the GRIB2 option better I do. What is the expected result in this case?

GeorgeGayno-NOAA commented:
@jordanschnell - this PR has not been updated in a while. There is more work required if you want to merge it. What do you want to do?

jordanschnell commented:
@GeorgeGayno-NOAA - We would like to merge it, however, I was not sure what else is required. Your question to Jeff and Larissa went unanswered and I am unsure of what the issue is. It seems like there may be other issues though (https://github.com/ufs-community/UFS_UTILS/issues/851)

LarissaReames-NOAA commented:
I think what @jordanschnell needs to do is create his own var mapping file based on the GSD var mapping file he's tested with. I don't think we should be modifying a file we've long used in the current form, especially when this doesn't reflect some operational change. This way there won't be any failures in the regression tests either.

How you'd choose the option for missing variables really depends. We never implemented any behavior to not write an array to file if "skip" is chosen and the variable is unavailable in the grib2 file. I believe the philosophy was always that, given a certain weather model SDF, we always wanted the atm init files to have certain arrays in them regardless of what data is available in the grib2 files used to create them.  If I remember correctly, we use "skip" for those variables that might be missing data at certain levels, like RH in some GFS grib2 files, so that a few missing levels wouldn't cause a failure. If you plan on a variable possible being completely missing from a grib2 file but don't want that to cause a failure, I'd suggest using set_to_fill. 

So, in that sort of paradigm, one might expect any weather model run utilizing HRRR Smoke to also require its own unique SDF, so it would make sense to have a separate variable mapping file. In that mapping file you could indeed set the missing behavior to be "stop" to make sure a user is initializing that weather model SDF with all the desired smoke tracers.

jordanschnell committed:
Created seperate file that contains smoke variable and removed smoke from the original varmap

GeorgeGayno-NOAA commented:
@jordanschnell - be sure to merge updates from develop to your branch.

jordanschnell commented:
> @jordanschnell - be sure to merge updates from develop to your branch.

@GeorgeGayno-NOAA, is that just:

`git checkout rap_icbc`
`git merge develop`

Then do I need to `git push`?

GeorgeGayno-NOAA commented:
> > @jordanschnell - be sure to merge updates from develop to your branch.
> 
> @GeorgeGayno-NOAA, is that just:
> 
> `git checkout rap_icbc` `git merge develop`
> 
> Then do I need to `git push`?

That will work if 'develop' in your fork is up-to-date with the authoritative fork. See our wiki page for how I do it (last two bullets). 

https://github.com/ufs-community/UFS_UTILS/wiki/3.-Checking-out-code-and-making-changes 

I like to do things from the command line. But, I would guess you can also do these updates from the web browser.

GeorgeGayno-NOAA commented:
@MichaelLueken - do you want to run your own tests?

GeorgeGayno-NOAA commented:
@jordanschnell - since this is a new configuration of chgres_cube, I think it would be a good idea to add a consistency test. I can help with this.

MichaelLueken commented:
@GeorgeGayno-NOAA - Sure, I should be able to run a test in the morning.  Thanks!

GeorgeGayno-NOAA commented:
After I checked out your fork, I did the following:

- git remote add upstream git@github.com:ufs-community/UFS_UTILS.git
- git remote update
- git remote -v
```
origin  git@github.com:jordanschnell/UFS_UTILS.git (fetch)
origin  git@github.com:jordanschnell/UFS_UTILS.git (push)
upstream        git@github.com:ufs-community/UFS_UTILS.git (fetch)
upstream        git@github.com:ufs-community/UFS_UTILS.git (push)
```
- git checkout upstream/develop
- git checkout develop
- git merge --no-commit upstream/develop
- git commit
- git push origin develop (I did not do this step as I don't want to push to your fork)

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I received an error: <img alt="Screenshot 2024-10-03 at 5 13 11 PM" width="686" src="https://private-user-images.githubusercontent.com/46031459/373416568-8ad24440-ac42-4b6a-8c08-340fb09abc17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgwNDU1ODQsIm5iZiI6MTcyODA0NTI4NCwicGF0aCI6Ii80NjAzMTQ1OS8zNzM0MTY1NjgtOGFkMjQ0NDAtYWM0Mi00YjZhLThjMDgtMzQwZmIwOWFiYzE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA0VDEyMzQ0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5YjI0MjIwZWUyNDY5M2I4Y2EyNDYyZjk4YTFiNGRmZjA2OGZjNjBjMzhkYjRkNWU5ODFjYTVmYmEwMjA3ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cUM12oAckJ5g6TFfV5BgAp8HfQBdhBr9lLtdRaiT6B8">
> 
> here were my commands: <img alt="Screenshot 2024-10-03 at 5 13 47 PM" width="467" src="https://private-user-images.githubusercontent.com/46031459/373416653-b5fad664-7e1f-47ee-bfa2-942fa892fcfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgwNDU1ODQsIm5iZiI6MTcyODA0NTI4NCwicGF0aCI6Ii80NjAzMTQ1OS8zNzM0MTY2NTMtYjVmYWQ2NjQtN2UxZi00N2VlLWJmYTItOTQyZmE4OTJmY2ZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA0VDEyMzQ0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3YTNiZjQ3NjMyMGI5MDZhODc5N2Y5MjY0NjI5NTg2YjE5OWFkZTEyZDgyNGYzN2IzMWExZGE0ZTc1ZDVhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8dd4NuSjtoVfnw0972vOZe3xpR-Mf9l2G1pXkpmSgv0">
> 
> and branches:
> 
> <img alt="Screenshot 2024-10-03 at 5 14 11 PM" width="304" src="https://private-user-images.githubusercontent.com/46031459/373416737-e131da94-a456-46c2-849a-677c21dde115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgwNDU1ODQsIm5iZiI6MTcyODA0NTI4NCwicGF0aCI6Ii80NjAzMTQ1OS8zNzM0MTY3MzctZTEzMWRhOTQtYTQ1Ni00NmMyLTg0OWEtNjc3YzIxZGRlMTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA0VDEyMzQ0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhYjQ1ODdmMGIyMzFkMjdjMzlmMTg0YmM2NzhmMjU4MGU2NDI4M2MxMWYzNmFiOGE4YjdkM2JiZWJiNmRlZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.G9rjXgEc9-OV5IhqIclwNIZXAW23ADlB0IqlTckLS6A">
> From the browser, would I "Sync fork"?

Don't clone using 'http', instead use, for example: `git@github.com:ufs-community/UFS_UTILS.git`

DavidHuber-NOAA reviewed (approved):
The math works out.  Thanks @jordanschnell!

MichaelLueken commented:
@GeorgeGayno-NOAA -

The SRW-SD WE2E test successfully passed:
```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used
----------------------------------------------------------------------------------------------------
smoke_dust_grid_RRFS_CONUS_3km_suite_HRRR_gf_20241004134527        COMPLETE            1746.09
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE            1746.09
```
I can also confirm that the smoke data from the ICs/LBCs were read into the weather model:

Smoke is being set to a value other than default (1E-12) -
`smoke  5.4751575E-02  7.9282990E-04  8.9739502E-04`

The fundamental WE2E tests were also run and all passed:
```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used 
----------------------------------------------------------------------------------------------------
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta_2  COMPLETE              11.56
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2_20241  COMPLETE               9.03
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8_plot  COMPLETE              25.11
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR_2024100  COMPLETE              43.06
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0_20241004171  COMPLETE              26.77
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16_2024100417143  COMPLETE              43.47
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE             159.00
```
With the original changes in `GSDphys_var_map.txt`, the `grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR` fundamental test failed.  This indicates that the test is pulling the unaltered `GSDphys_var_map.txt` file and successfully running.

The comprehensive WE2E tests have also successfully passed without issue.

This update is running correctly in the SRW App.

jordanschnell commented:
@GeorgeGayno-NOAA @DavidHuber-NOAA my develop is now up to date. Thank you for the help!

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA @DavidHuber-NOAA my develop is now up to date. Thank you for the help!

Great. Now, merge the updates from 'develop' to your branch. See bottom of this page:
https://github.com/ufs-community/UFS_UTILS/wiki/3.-Checking-out-code-and-making-changes

jordanschnell committed:
Merge branch 'ufs-community:develop' into rap_icbc

jordanschnell commented:
@GeorgeGayno-NOAA I think I was successful

GeorgeGayno-NOAA commented:
@jordanschnell - do you have simple case that I can convert into a regression test? The test would be added here:
https://github.com/ufs-community/UFS_UTILS/tree/develop/reg_tests/chgres_cube

jordanschnell commented:
> @jordanschnell - do you have simple case that I can convert into a regression test? The test would be added here: https://github.com/ufs-community/UFS_UTILS/tree/develop/reg_tests/chgres_cube

I think this one could be duplicated: reg_tests/chgres_cube/13km.conus.rap.grib2.sh

The only change needed would be the varmap.

GeorgeGayno-NOAA commented:
> > @jordanschnell - do you have simple case that I can convert into a regression test? The test would be added here: https://github.com/ufs-community/UFS_UTILS/tree/develop/reg_tests/chgres_cube
> 
> I think this one could be duplicated: reg_tests/chgres_cube/13km.conus.rap.grib2.sh
> 
> The only change needed would be the varmap.

Can you try to create the test and run it from the driver script for your machine (I will help)? Does it need different input GRIB data from the 13km.conus.rap.grib2.sh test? I can add the input data to the baseline directory.

jordanschnell commented:
> > > @jordanschnell - do you have simple case that I can convert into a regression test? The test would be added here: https://github.com/ufs-community/UFS_UTILS/tree/develop/reg_tests/chgres_cube
> > 
> > 
> > I think this one could be duplicated: reg_tests/chgres_cube/13km.conus.rap.grib2.sh
> > The only change needed would be the varmap.
> 
> Can you try to create the test and run it from the driver script for your machine (I will help)? Does it need different input GRIB data from the 13km.conus.rap.grib2.sh test? I can add the input data to the baseline directory.

smoke/massden is in the grib2 input data so it can be used.  I'll set it up and let you know if I have issues.

GeorgeGayno-NOAA commented:
I made one change to the 13km.conus.rap.grib2 test:

```
diff --git a/reg_tests/chgres_cube/13km.conus.rap.grib2.sh b/reg_tests/chgres_cube/13km.conus.rap.grib2.sh
index 4da2468d..c0338b12 100755
--- a/reg_tests/chgres_cube/13km.conus.rap.grib2.sh
+++ b/reg_tests/chgres_cube/13km.conus.rap.grib2.sh
@@ -20,7 +20,7 @@ export COMIN=${HOMEreg}/input_data/rap.grib2

 export GRIB2_FILE_INPUT=1921221000900
 export VCOORD_FILE=${HOMEufs}/fix/am/global_hyblev.l64.txt
-export VARMAP_FILE=${HOMEufs}/parm/varmap_tables/GSDphys_var_map.txt
+export VARMAP_FILE=${HOMEufs}/parm/varmap_tables/GSDphys_smoke_var_map.txt
 export INPUT_TYPE='grib2'
 export CONVERT_NST=".false."

```

It ran to completion, and I see a smoke record in the output. Would this be a valid test? If so, make a copy of the 13km.conus.rap.grib2.sh script, give it a descriptive name, then add its call from the driver script. Try it on one machine first. Check the output. If it is correct, I will add it to the baseline directory.

jordanschnell committed:
add regtests for smoke

jordanschnell committed:
update reg

jordanschnell commented:
Smoke is in the output (/lfs5/BMC/rtwbl/rap-chem/tmp/regtest_out/reg-tests/chgres-cube/13km_conus_rap-smoke_grib2), but the regtest failed because it is not in the baseline.

GeorgeGayno-NOAA commented:
> Smoke is in the output (/lfs5/BMC/rtwbl/rap-chem/tmp/regtest_out/reg-tests/chgres-cube/13km_conus_rap-smoke_grib2), but the regtest failed because it is not in the baseline.

I created a baseline directory. Update the `13km.conus.rap-smoke.grib2.sh` script as follows and try again:

```
diff --git a/reg_tests/chgres_cube/13km.conus.rap-smoke.grib2.sh b/reg_tests/chgres_cube/13km.conus.rap-smoke.grib2.sh
index e80e0579..2a7bfa19 100755
--- a/reg_tests/chgres_cube/13km.conus.rap-smoke.grib2.sh
+++ b/reg_tests/chgres_cube/13km.conus.rap-smoke.grib2.sh
@@ -81,7 +81,7 @@ for files in *.nc
 do
   if [ -f $files ]; then
     echo CHECK $files
-    $NCCMP -dmfqS $files $HOMEreg/baseline_data/13km_conus_rap_grib2/$files
+    $NCCMP -dmfqS $files $HOMEreg/baseline_data/13km_conus_rap-smoke_grib2/$files
     iret=$?
     if [ $iret -ne 0 ]; then
```

GeorgeGayno-NOAA commented:
I created baseline directories on all Tier 1 machines.

jordanschnell commented:
regression test: passed: /mnt/lfs5/BMC/rtwbl/rap-chem/src/RRFS/rrfs-workflow_rap_icbc_regtest/sorc/UFS_UTILS/reg_tests/chgres_cube/consistency.log14

![Screenshot 2024-10-17 at 5 50 15 PM](https://github.com/user-attachments/assets/633da160-2c4a-41da-8b10-962fce345adb)

GeorgeGayno-NOAA commented:
> regression test: passed: /mnt/lfs5/BMC/rtwbl/rap-chem/src/RRFS/rrfs-workflow_rap_icbc_regtest/sorc/UFS_UTILS/reg_tests/chgres_cube/consistency.log14
> 
> ![Screenshot 2024-10-17 at 5 50 15 PM](https://private-user-images.githubusercontent.com/46031459/377643923-633da160-2c4a-41da-8b10-962fce345adb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyNTMyNjgsIm5iZiI6MTcyOTI1Mjk2OCwicGF0aCI6Ii80NjAzMTQ1OS8zNzc2NDM5MjMtNjMzZGExNjAtMmM0YS00MWRhLThiMTAtOTYyZmNlMzQ1YWRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDEyMDI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlMThjNmVmYTVmMWYyZDc2Y2E3ZjU5ZTA5ZTNjNzNjNjVkMmQxYmE5MWNlOTMxNzA1MzE5NzY2MWMwZDE1MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KSHygzjfcL4bw12qwC44auZ27_McOwMgi2B6FCQv7sw)

Great. Do you have access to the other Tier 1 machines? Hera, Orion, Hercules and WCOSS2? If not, I can update the driver scripts on those machines and run the tests.

GeorgeGayno-NOAA commented on reg_tests/chgres_cube/13km.conus.rap-smoke.grib2.sh:
```diff
@@ -0,0 +1,102 @@
+#!/bin/bash
+
+#-----------------------------------------------------------------------------
+# Invoke chgres to create 13-km CONUS coldstart files using RAP GRIB2 data
```
I would add something to the description about this testing the 'smoke' option - what it is and when it is used.

GeorgeGayno-NOAA replied:
Resolved at a4de98b.

GeorgeGayno-NOAA commented:
Because you are adding a varmap file, this section of the documentation may need some updates:
https://noaa-emcufs-utils.readthedocs.io/en/latest/ufs_utils.html#variable-mapping-varmap-table

I can help with that.

jordanschnell commented:
@GeorgeGayno-NOAA I have access to Hera, but not the others.  so, these are my remaining tasks?

1) Update driver.hera script and test there?
2) Add comments to regtest
3) Add documentation

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I have access to Hera, but not the others. so, these are my remaining tasks?
> 
> 1. Update driver.hera script and test there?
> 2. Add comments to regtest
> 3. Add documentation

And do as many of the tests described above in "TESTS CONDUCTED". I can do the tests on the Herc, Orion and WCOSS2.

jordanschnell committed:
add hera regtest

jordanschnell commented:
Regtest on Hera passed: /scratch1/BMC/wrf-chem/Jordan/RRFS-SD/src/PRs/rap_icbc/rrfs-workflow/sorc/UFS_UTILS/reg_tests/chgres_cube/consistency.log14

GeorgeGayno-NOAA commented:
@jordanschnell - I updated and tested the regression test driver scripts on Hercules, Orion and WCOSS2. I placed them on Hera here: /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.schnell/reg_tests/chgres_cube

They have ".updated" in their filename.

jordanschnell commented:
> @jordanschnell - I updated and tested the regression test driver scripts on Hercules, Orion and WCOSS2. I placed them on Hera here: /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.schnell/reg_tests/chgres_cube
> 
> They have ".updated" in their filename.

Great thank you! So commit these?

jordanschnell committed:
Update drivers for hercules, wcoss2, and orion

jordanschnell commented:
> > Because you are adding a varmap file, this section of the documentation may need some updates: https://noaa-emcufs-utils.readthedocs.io/en/latest/ufs_utils.html#variable-mapping-varmap-table
> > I can help with that.
> 
> @jordanschnell - Once you update the documentation, I will approve and merge this PR.

Great - how exactly do I update this?

As for specifically what to add, are you thinking I add a note under the Column 1, something to the effect of:

*** If smoke from RAP/HRRR are desired, one must select the GSDphys_smoke_var_map.txt varmap file which includes

massden - smoke mass mixing ratio

***

GeorgeGayno-NOAA commented:
> > > Because you are adding a varmap file, this section of the documentation may need some updates: https://noaa-emcufs-utils.readthedocs.io/en/latest/ufs_utils.html#variable-mapping-varmap-table
> > > I can help with that.
> > 
> > 
> > @jordanschnell - Once you update the documentation, I will approve and merge this PR.
> 
> Great - how exactly do I update this?
> 
> As for specifically what to add, are you thinking I add a note under the Column 1, something to the effect of:
> 
> *** If smoke from RAP/HRRR is desired, one must select the GSDphys_smoke_var_map.txt varmap file which includes
> 
> massden - smoke mass mixing ratio

I set up readthedocs for your branch here:
https://schnel.readthedocs.io/en/rap_icbc/ufs_utils.html#variable-mapping-varmap-table

You will edit the ./docs/source/chgres_cube.rst file. When you push your branch to your fork, the new documentation should automatically update.

Yes, I would add the definition of massden to the list. Try that first and see if everything works. Then, we can add additional instructions for users.

jordanschnell committed:
doc test

jordanschnell committed:
more

jordanschnell commented:
Commit is pushed, but I don't see it in the docs branch you created

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA - Commit is pushed, but I don't see it in the docs branch you created

Looks like I have to run it manually. Take a look now. The formatting does not look right.

jordanschnell committed:
format fix attempt

GeorgeGayno-NOAA commented:
> > @GeorgeGayno-NOAA - Commit is pushed, but I don't see it in the docs branch you created
> 
> Looks like I have to run it manually. Take a look now. The formatting does not look right.

I would just add the definition of massden.

jordanschnell committed:
minimum docs

jordanschnell commented:
> > > @GeorgeGayno-NOAA - Commit is pushed, but I don't see it in the docs branch you created
> > 
> > 
> > Looks like I have to run it manually. Take a look now. The formatting does not look right.
> 
> I would just add the definition of massden.

ok hopefully that worked

<img width="540" alt="Screenshot 2024-11-04 at 3 05 14 PM" src="https://github.com/user-attachments/assets/f90e5951-60f0-4582-bb20-1a83f82eaae1">

jordanschnell committed:
remove note that only GFS physics suite is supported

jordanschnell commented:
> The format looks good.
> 
> One more favor, can you remove the line: "(Note, only the GFS physics suite is currently supported.)" and add "Each file contains the following **columns**:"

Should be good to go!

GeorgeGayno-NOAA commented:
PR #993 will be merged before this one. When that happens, you will need to merge those updates to your branch. PR #993 updates another program. So, there won't be any conflicts.

GeorgeGayno-NOAA commented:
@jordanschnell - please merge the latest changes from 'develop' to your branch. I will then do a few final checks, then merge your changes.

jordanschnell committed:
Merge branch 'ufs-community:develop' into rap_icbc
